---
title: "R Notebook"
output: html_notebook
---
#Installer Blast+ https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/
#Télécharger database 16s de NCBI https://ftp.ncbi.nlm.nih.gov/blast/db/ (16S_ribosomal_RNA...)
#Hahsler M, Anurag N (2024). rBLAST: R Interface for the Basic Local Alignment Search Tool. R package version 1.0.0, https://github.com/mhahsler/rBLAST.

```{r, eval=FALSE, packages mais il doit en manquer, voir avec les library}
install.packages(c("VennDiagram", "VennCounts"))
install.packages("readxl")
```

```{r library, include=FALSE}
library(readxl)
library(openxlsx)
library(VennDiagram)
library(DECIPHER)
library(openxlsx)
library(rmarkdown)
library(knitr)
library(phyloseq)
library(dada2)
library(DECIPHER)
library(phangorn)
library(ggplot2)
library(gridExtra)
library(shiny)
library(miniUI)
library(caret)
library(pls)
library(e1071)
library(ggplot2)
library(randomForest)
library(dplyr)
library(ggrepel)
#library(nlme)
library(devtools)
library(reshape2)
library(PMA)
#library(structSSI)
library(ade4)
library(ggnetwork)
library(intergraph)
library(scales)
library(genefilter)
library(impute)
library(phyloseqGraphTest)
library(Biostrings)
library(RSQLite)
library(parallel)
library(ape)
library(taxonomizr)
library('rBLAST')
library(reshape2)
library(ggvenn)
library(cowplot)
library(patchwork)
```

```{r, eval=FALSE, créer la data base, prévoire 2-3h et 80go de stockage, NE LANCER qu'une fois}
libdir='data'
dir.create(libdir)
setwd(libdir)
getNamesAndNodes()
getAccession2taxid(types=c('nucl_gb'))
getAccession2taxid()
system("gunzip *.gz")
read.accession2taxid(list.files('.','accession2taxid'),'accessionTaxa.sql')
print(paste('taxonomizr database built and located at', getwd(), sep=' '))
```

```{r, Ne lancer que si la commande suivante ne fonctionne pas en spécifiant le path de blast+}
Sys.setenv(PATH = paste(Sys.getenv("PATH"), "C:/Program Files/NCBI/blast-2.15.0+/bin", sep= .Platform$path.sep))
```

```{r}
#BLAST query
file_list <- list.files('/Users/amrozins/Documents/workflow/Sequences/Manon B/trim', pattern = "\\.fa$", full.names = TRUE)

dna <- readDNAStringSet(file_list, format='fasta')


bl <- blast(db="/Users/amrozins/Documents/workflow/16S_ribosomal_RNA/16S_ribosomal_RNA", type = "blastn")

#Run BLAST query
fmt <- "qseqid saccver pident length evalue qcovs"
#When not provided, the default value is: 'qaccver saccver pident length mismatch gapopen qstart qend sstart send evalue bitscore'
#voir rBLAST help
cl_dna <- predict(bl, dna, custom_format = fmt)

#affiche les premiers 5 hits
cl_dna[1:5,]
```

```{r}
accid = as.character(cl_dna$saccver)
```

```{r}
taxaNodes<-read.nodes.sql("/Users/amrozins/Documents/workflow/data/nodes.dmp")
taxaNames<-read.names.sql("/Users/amrozins/Documents/workflow/data/names.dmp")
```

```{r}
#takes accession number and gets the taxonomic ID
ids<-accessionToTaxa(accid, '/Users/amrozins/Documents/workflow/data/accessionTaxa.sql')

#taxlist displays the taxonomic names from each ID #
taxlist=getTaxonomy(ids, taxaNodes, taxaNames)
```

```{r, Résultats Totaux BarnenezEau}
cltax=cbind(cl_dna,taxlist) #bind BLAST hits and taxonomy table
colnames(cltax)
cltax
```

```{r, top 5}
# Tri des résultats par qseqid et pident décroissants
cltax_sorted <- cltax %>% arrange(qseqid, desc(pident))

# Sélection des 5 premiers résultats pour chaque qseqid
cltax5 <- cltax_sorted %>% group_by(qseqid) %>% slice_head(n = 5)
print(cltax5)
```

```{r, max % identité}
#Ne prend pour chaque fasta que le résultat avec le max de % identity (pident)
filtered_cltax <- cltax_sorted %>%
  group_by(qseqid) %>%
  filter(pident == max(pident))

filtered_cltax
```


```{r}
data <- subset(filtered_cltax, 
                  select=c(qseqid, superkingdom, phylum, class, order, family, genus, species, pident, evalue, qcovs, length))
chemin_du_dossier <- "/Users/amrozins/Documents/workflow/Sequences/Manon B/trim"

fasta <- function(qseqid, chemin_du_dossier) {
    pattern <- paste0("^", gsub("\\.ab1$", "", qseqid), ".*\\.fa$")
    fichiers <- list.files(chemin_du_dossier, pattern = pattern, full.names = TRUE, ignore.case = TRUE)
    
    if (length(fichiers) > 0) {
        seqs <- readDNAStringSet(fichiers[1], format = "fasta")
        return(as.character(seqs))
    } else {
        cat("Aucun fichier correspondant trouvé pour :", qseqid, "\n")
        return(NA)
    }
}

data$sequence <- sapply(data$qseqid, fasta, chemin_du_dossier)
data$longueur_sequence_pb <- sapply(data$sequence, function(seq) ifelse(is.na(seq), NA, nchar(seq)))
data <- data[, c(names(data)[names(data) != "sequence"], "sequence")]
print(data)
```



```{r, stats sur les sequences}
# Calculer les moyennes et les écarts types
mean_length <- mean(data$length, na.rm = TRUE)
mean_qcovs <- mean(data$qcovs, na.rm = TRUE)
mean_pident <- mean(data$pident, na.rm = TRUE)
sd_length <- sd(data$length, na.rm = TRUE)
sd_qcovs<- sd(data$qcovs, na.rm = TRUE)
sd_pident <- sd(data$pident, na.rm = TRUE)

# Afficher les moyennes et les écarts types pour all_filtered_dataVerveur
cat("Moyennes et écarts types ")
cat("length: ", mean_length, " (", sd_length, ")\n")
cat("qcovs: ", mean_qcovs, " (", sd_qcovs, ")\n")
cat("pident: ", mean_pident, " (", sd_pident, ")\n\n")
```


```{r, data en supprimant les doublon, garde la première occurence}
# Filtrage des colonnes nécessaires
datatrim <- subset(filtered_cltax, 
                  select=c(qseqid, superkingdom, phylum, class, order, family, genus, species, pident, evalue, qcovs, length))

# Définition du chemin du dossier
chemin_du_dossier <- "/Users/amrozins/Documents/workflow/Sequences/Manon B/trim"

# Fonction pour récupérer les séquences FASTA
fasta <- function(qseqid, chemin_du_dossier) {
    pattern <- paste0("^", gsub("\\.ab1$", "", qseqid), ".*\\.fa$")
    fichiers <- list.files(chemin_du_dossier, pattern = pattern, full.names = TRUE, ignore.case = TRUE)
    
    if (length(fichiers) > 0) {
        seqs <- readDNAStringSet(fichiers[1], format = "fasta")
        return(as.character(seqs))
    } else {
        cat("Aucun fichier correspondant trouvé pour :", qseqid, "\n")
        return(NA)
    }
}

# Ajout des séquences et de leur longueur
datatrim$sequence <- sapply(datatrim$qseqid, fasta, chemin_du_dossier)
datatrim$longueur_sequence_pb <- sapply(datatrim$sequence, function(seq) ifelse(is.na(seq), NA, nchar(seq)))

# Réarrangement des colonnes
datatrim <- datatrim[, c(names(data)[names(data) != "sequence"], "sequence")]

# Suppression des doublons basés sur la colonne qseqid
datatrim <- data[!duplicated(data$qseqid), ]

# Affichage du résultat
print(datatrim)

```

```{r, stats sur datatrim}
# Calculer les moyennes et les écarts types
mean_length <- mean(datatrim$length, na.rm = TRUE)
mean_qcovs <- mean(datatrim$qcovs, na.rm = TRUE)
mean_pident <- mean(datatrim$pident, na.rm = TRUE)
sd_length <- sd(datatrim$length, na.rm = TRUE)
sd_qcovs<- sd(datatrim$qcovs, na.rm = TRUE)
sd_pident <- sd(datatrim$pident, na.rm = TRUE)

# Afficher les moyennes et les écarts types pour all_filtered_dataVerveur
cat("Moyennes et écarts types
    ")
cat("length: ", mean_length, " (", sd_length, ")\n")
cat("qcovs: ", mean_qcovs, " (", sd_qcovs, ")\n")
cat("pident: ", mean_pident, " (", sd_pident, ")\n\n")
```

```{r}
# Tableau de comptage des genres
genres <- table(datatrim$genus)

# Compter le nombre total de genres
total_genres <- length(unique(datatrim$genus))

# Créez le graphique à barres
bar_plot <- ggplot(data.frame(Genus = names(genres), Count = as.numeric(genres)), 
                   aes(x = reorder(Genus, -Count), y = Count)) +
  geom_bar(stat = "identity", fill = "#145C9E") +
  geom_text(aes(label = Count), vjust = -0.5, color = "black", size = 3.5) +  # Ajouter les valeurs au-dessus de chaque barre
  labs(title = "Genres retrouvés", x = "Genres", y = "Nombre d'e séquences'isolats") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_y_continuous(breaks = seq(0, 10, by = 1), limits = c(0, 10)) +  # Régler les valeurs de l'axe des ordonnées
  annotate("text", x = 15, y = 9.5, label = paste("Total de séquences :", nrow(datatrim)), hjust = 1, vjust = 1) +
  annotate("text", x = 15, y = 9, label = paste("Nombre total de genres :", total_genres), hjust = 1, vjust = 0.95) # Ajouter le nombre total de genres

print(bar_plot)
```



```{r, classes}
class_counts <- table(datatrim$class)

bar_plot_classes <- ggplot(data.frame(Class = names(class_counts), Count = as.numeric(class_counts)), 
                              aes(x = reorder(Class, -Count), y = Count)) +
  geom_bar(stat = "identity", fill = "#59C9A5") +
  geom_text(aes(label = Count), position = position_dodge(width = 0.9), vjust = -0.5, color = "black", size = 3.5) +
  labs(title = "Classes retrouvées", x = "Class", y = "Nombre d'isolats") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+
  # Ajout de l'annotation pour le nombre total de séquences
  annotate("text", x = Inf, y = -Inf, label = paste("Total d'isolats :", nrow(datatrim)), hjust = 1, vjust = 0)

print(bar_plot_classes)
```



```{r}
write.csv(data, "/Users/amrozins/Documents/workflow/Résultats/rBlastARN/data.csv", row.names = FALSE)
write.csv(datatrim, "/Users/amrozins/Documents/workflow/Résultats/rBlastARN/datatrim.csv", row.names = FALSE)
ggsave("/Users/amrozins/Documents/workflow/Résultats/rBlastARN/bar_plot.png", bar_plot, width = 10, height = 6, units = "in")
ggsave("/Users/amrozins/Documents/workflow/Résultats/rBlastARN/bar_plot_classes.png", bar_plot_classes, width = 10, height = 6, units = "in")
```

